README.md
# IEEE VTS Motor Vehicles Challenge 2023: A Multi-physical Benchmark Problem for Next Generation Energy Management Algorithms

## Introduction

The competitors are invited to develop the energy management algorithm (EMA) for the vehicle (cf. Figure 1).
The EMA determines the operating conditions for the two energy storage devices and the three electric motors; minimization of the energy consumption and battery degradation are some  of the main goals of the EMA.

## Announcement

**Registration is closed** and the repository is in final state.

**Please follow the** [**submission guideline** for your proposed EMA using a private GITHUB repository as described](SUBGUIDE.md).  
Make sure **your EMA model can be executed by the "main.m" script**. This is crucial for the later (partial automated) assessment process of your EMA in the competition. 

Good luck! 

### Preliminary timeline

- EMA submission deadline ~ End of March 2023.
- [Paper submission deadline - 17th of April 2023.](https://events.vtsociety.org/vppc2023/)

### Release Notes

Please refer to [the change log](CHANGELOG.md).  
**Latest: 17th of Feb. 2023 - Final Release** (i.a. 30&nbsp;s preview horizon).  
**Remarks:** 
- Please take notice of [the discussion board, with a lot of solved Q&A](https://github.com/DLR-VSDC/IEEE-MVC-2023/discussions).
- Moreover, it is worthwhile to have a look at the [solved issues section](https://github.com/DLR-VSDC/IEEE-MVC-2023/issues?q=is%3Aissue).

## Important Links

- How to get ready in MATLAB/Simulink, please refer to [installation readme](Installation.md).
- For additional information please view presentation slides of the paper [*.pdf](./media/IEEE_VTS_MVC2023.pdf).
- The [submitted version of the publication, **incl. fixed errata**](https://elib.dlr.de/193953/) can be found in our elib repository.
- For information on the given tracks, please refer to [tracks readme](/Tracks/Readme.md).

### THE MVC 2023 Benchmark Problem

In Figure 1 the MVC 2023 challenge model and it's components are shown.  
The competitors are invited to develop the top red block, the energy management algorithm (EMA).

<img src="./media/EMA_block_diagram.png" style="width:3.3686in;height:3.29562in" />

Figure 1: Block diagram of the MVC 2023 benchmark problem.

## Bibliography

Brembeck, J.; de Castro, R.; Tobolar, J. & Ebrahimi, I.
IEEE VTS Motor Vehicles Challenge 2023: A Multi-physical Benchmark Problem for Next Generation Energy Management Algorithms 
*19th IEEE Vehicle Power and Propulsion Conference (VPPC)*, **2022** 

Link to [reference *.bib file](./media/Bre22_MVC2023.bib).

## License

Copyright © 2022-2023 DLR & UCM. The code is released under the [CC BY-NC 4.0 license](https://creativecommons.org/licenses/by-nc/4.0/legalcode). Link to [short summary of CC BY-NC 4.0 license](https://creativecommons.org/licenses/by-nc/4.0/). For attribution see also [license file](LICENSE.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [Final Release] - 2023-02-17

### Added
- "Exit Flag" indicator added to the matlab report. This value is 1, when the SoC safety constraints of both the battery and the FC are violated, 0 otherwise. 
- J_flag is added to the total performance metric to penalize when Exit Flag is 1.

### Changed
- Increased the preview horizon to 30 seconds with sampling time of 1 second. The signal under this name contains a vector of dimention 60x1, where the first 30 elements are the velocity and the remaining 30 elements are the curvature preview values. For the sake of convenience, selectors were added to this signal to extract the speed and curvature values on the baseline EMA subsystem.
- Accuracy of the performance metrics was increased to 8 digits. The Baseline-Performance mat file is updated accordingly.
- Increased "Maximum allowed lateral acceleration" to 10&nbsp;m/s2, see https://github.com/DLR-VSDC/IEEE-MVC-2023/issues/27.

### Fixed
- Fixed a typo on the battery's degradation unit. Correct unit: [Ah/cycle]
- [Updated submitted version of the publication, incl. all fixed errata](https://elib.dlr.de/193953), solves latest: https://github.com/DLR-VSDC/IEEE-MVC-2023/discussions/18

## [Beta 3] - 2023-01-13

### Changed
Changes on main.m script:
- Added FC SOC safety to stop the car if SoC_H2 tank goes below 1%.
- Increased minimum allowed FC SoC (param.fuelcell.SOCmin) to 10%.
- Reduced H2 tank size to 0.15 kg.
- Decreased battery's maximum allowed SoC to 80%.
- Decreased safety stop tolerance level of battery to 99% from 100.1%.
- Lowered maximum allowed temperature to 35 deg C.

Changes on Simulink model (testEnergyManagementAlgorithm.slx):
- Initial velocity of the vehicle is set to zero. This solves [discussion #11](https://github.com/DLR-VSDC/IEEE-MVC-2023/discussions/11)
- Safety stop condition have been added to the model/Performance Metrics, the one with yellow comment.
- Inside Baseline EMA block, few additional signals are added that competitors could have access to.
- Renamed H2_I to H2_content.
- Included P_battery directly instead of computing it from P=V*I.
- Removed some unnecessary scopes.


### Fixed

- Updated documentation of aging model - [Discussion #15](https://github.com/DLR-VSDC/IEEE-MVC-2023/discussions/15),   
refer to [updated chapter 3c of the paper](./media/MVC2023_chapter3C_update_Beta3.pdf)
- New FMU model `MVC2023_Hash_47c69bf5.fmu` changed battey voltage signal from OCV to the acutual measured value in the circuit.    
 Added some additional signals to the batterybus and rexbus. This solves [discussion #11](https://github.com/DLR-VSDC/IEEE-MVC-2023/discussions/11)

## [Beta 2] - 2023-01-03

### Changed

- New parameterization of the intial values, please refer to main.m - please mind that these might be changed in the final version
- [Extend state vector in the EMA #8](https://github.com/DLR-VSDC/IEEE-MVC-2023/issues/8)

### Fixed
- Revised model of the fuel cell - [Fuel Cell Model: wrong SI-units in H2 consumption #10](https://github.com/DLR-VSDC/IEEE-MVC-2023/issues/10):
	- New FMU model `MVC2023_Hash_ddadadfe.fmu` with revised fuel cell model
	- Corrected fuel cell consumption table (wrong units) and Modelica model, refer to [updated chapter 3e of the paper](./media/MVC2023_chapter3E_update_Beta2.pdf)


## [Beta 1] - 2022-12-16

### Added

- A new Performance metric is included, and it is called `J_tire`. This performance metric captures tire losses, to evaluate the torque vectoring capability of the designed EMA.

### Changed

- [Updated chapter 4c of the paper (Performance Metrics)](./media/MVC2023_chapter4C_update_Beta1.pdf).
- Track naming is updated according to their mission tracks (e.g. [FTP75.mat](ftp://ftp75.mat/), ECE15.mat, etc.).
- Some tracks are extended (repeated over and over) to challenge the energy management algorithms to drive the vehicle in more efficient way. (Some tracks may be longer than what the power source can deliver)
- Fuel Cell's H2 tank size has been reduced to 40 kg.
- After each run of the main script, a folder will be auto-generated on `/Reporting/Report_History` with the name of the track, date and time. This folder contains HTML report and the generated performance plots. This could help the participants to evaluate their progress.
- `J_engergy` is extended to the total energy, i.e. energy provided by the battery and the fuel cell and all the losses.
- The final score of the EMA is computed as the normalized metrics with respect to the Baseline Policy against each track. If one `J_x` is nonzero while baseline gives zero, the overall `J` will be penalized by a large number.
- Competitors are invited to improve the EMA such that the overall performance metric goes below 1 (`J_tot = 1` is when the designed EMA works exactly as the baseline EMA, `J_tot > 1` works worse and `J_tot < 1` better).

### Fixed

## [Alpha] - 2022-11-02

### Added

- Initial version of the IEEE MVC 2023 material

## License
Copyright © 2022-2023 DLR & UCM. The code is released under the [CC BY-NC 4.0 license](https://creativecommons.org/licenses/by-nc/4.0/legalcode). Link to [short summary of CC BY-NC 4.0 license](https://creativecommons.org/licenses/by-nc/4.0/)


Main.m
% IEEE VTS Motor Vehicles Challenge 2023
% University of California, Merced
% Main Script
%----------------------------------------------------
clc;clear;close all;
addpath(genpath('Reporting'))
addpath('Functions')
addpath('Tracks')
% add the folder to the MATLAB path
addpath(fullfile(pwd, 'FMIKit-Simulink-3.0'))
% initialize FMI Kit
FMIKit.initialize()

%% Initialization
Ts_solver   = 0.01;                                                         % Solver maximum sampling time
Ts_sample   = 0.01;                                                         % Sampling of tracks and outputs

% Track selection: For more information on the tracks, please refer to the Reamdme.md file on the "Tracks" folder.
list        = { 'Track No.  1: FTP75',...
                'Track No.  2: ECE 15',...
                'Track No.  3: Artemis Urban',...
                'Track No.  4: Artemis Rural',...
                'Track No.  5: ViresRuralRoadDescent1',...
                'Track No.  6: ViresRuralRoadDescent2',...
                'Track No.  7: NonOptMunich1',...
                'Track No.  8: OptMunich1',...
                'Track No.  9: NonOptMunich2' ...
                'Track No. 10: OptMunich2'};
[ind,tf]    = listdlg('PromptString',{'Please select a track',...
            'Please note: Only one can be selected at a time'},'SelectionMode','single','ListString',list,'ListSize',[300 200]);

N_pre       = 30;                                                           % Preview window size

% Generate inputs to the Simulink model
[Track,urlReport,urlHTML,urlPublish,FolderName,...
    preview,refSpeed,refCurv,Tend,T_s] = trackSelect(ind,N_pre);

fprintf('Simulation of the Mission: %s',Track)
fprintf('\n\nInitializing the Simulink model:\n')
%% System Parameters
% The following parameters are not allowed to be changed.
param.fuelcell.Imax         = 40;                                           % [A] maximum Fuel cell current
param.fuelcell.SOCmax       = 1;                                            % [-] maximum allowed SoC
param.fuelcell.SOCmin       = 0.1;                                          % [-] minimum allowed SoC

param.fuelcell.H2_tank_level= 0.15;                                         % [kg] H2 tank initial level value

param.bat.SOCmax            = 0.80;                                         % [-] maximum allowed SoC
param.bat.SOCmin            = 0.20;                                         % [-] minimum allowed SoC
param.bat.SOC_delta         = 0.05;                                         % [-] SOC for derating
param.bat.SOC_init          = 0.65;                                         % [-] vehicle battery start SoC

% safety battery/FC SoC values
param.fuelcell.SOCmin_stop  = 0.01;                                         % [-] vehicle shutdown if FC SoC become lower than this value
param.fuelcell.SOCmax_stop  = 1;                                            % [-] vehicle shutdown if FC SoC become higher than this value

param.bat.SOCmin_stop       = 0.01;                                         % [-] vehicle shutdown if battery SoC become lower than this value
param.bat.SOCmax_stop       = 0.99;                                         % [-] vehicle shutdown if battery SoC become higher than this value
param.bat.Tmax              = 35;                                           % [deg C] maximum allowed temperature

% Baseline EMA parameters
param.EMAbaseline.kfc       = 0.8;                                          % fuel cell distribution ratio
param.EMAbaseline.ktv       = 0.5;                                          % torque vectoring ratio
param.EMAbaseline.a_ad      = 0.5;                                          % front/rear ratio

%% Main simulation
fprintf('Please wait until the simulation is completed...\n\n')
tic;
out = sim('testEnergyManagementAlgorithm.slx','StopTime',num2str(Tend));
sim_time = toc;
fprintf('Success!\t Total simulation time is: \t %d seconds\n',sim_time)
fprintf('Please check the auto-generated html report for further evaluation of your designed EMA.\n')

%% Reporting (The following lines must remain unchanged)
Results;
options.showCode = false;
publish(urlReport,options);
web(urlPublish)
